Using `_id` in aggregate functions throws an exception

### Elasticsearch Version

9.2.0-SNAPSHOT, but also on 9.1.2

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

N/A (runs on ECH)

### Problem Description

When using the `_id` field in an aggregate function, ES throws the following exception:

> java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null

Simple example query resulting in an exception:
````
FROM metrics-*.otel-* METADATA _id
  | STATS c = COUNT(_id) BY service.name
````

This behavior appears to be indeterministic because it sometimes succeeds, but most of the time not. 

The same happens also with other aggregate functions (e.g. `... SAMPLE(_id,10) ...`).

Changing the above query by simply exchanging `_id` with `*` resolves the issue:
````
FROM metrics-*.otel-* METADATA _id
  | STATS c = COUNT(*) BY service.name
````




### Steps to Reproduce

1. Go to https://otel.demo.elastic.co/ 
2. Go to `Try ES|QL` in Discover
3. Select a time range of 1 hour
4. Run the following query:
```
FROM metrics-*.otel-* METADATA _id
| STATS c = COUNT(_id) BY service.name
```


<details>
  <summary>
<h3> (expand to see the) Error Stacktrace from Dev Tools </h3>
</summary>

```
org.elasticsearch.ElasticsearchException$1: Cannot invoke "String.length()" because "v" is null
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:798)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.ElasticsearchException.generateFailureXContent(ElasticsearchException.java:712)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.rest.RestResponse.build(RestResponse.java:202)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.rest.RestResponse.<init>(RestResponse.java:161)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.rest.RestResponse.<init>(RestResponse.java:122)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.rest.action.RestActionListener.onFailure(RestActionListener.java:54)
	at org.elasticsearch.xpack.esql.action.EsqlResponseListener.lambda$wrapWithLogging$1(EsqlResponseListener.java:191)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:266)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.rest.action.RestCancellableNodeClient$1.onFailure(RestCancellableNodeClient.java:97)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.tasks.TaskManager$1.onFailure(TaskManager.java:240)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onFailure(ActionListenerImplementations.java:360)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:40)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.xpack.esql.core.async.AsyncTaskManagementService.lambda$wrapStoringListener$5(AsyncTaskManagementService.java:258)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:266)
	at org.elasticsearch.xpack.esql.plugin.TransportEsqlQueryAction.lambda$innerExecute$9(TransportEsqlQueryAction.java:301)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:266)
	at org.elasticsearch.xpack.esql.execution.PlanExecutor.onQueryFailure(PlanExecutor.java:131)
	at org.elasticsearch.xpack.esql.execution.PlanExecutor.lambda$esql$1(PlanExecutor.java:106)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:266)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener$FailureResult.complete(SubscribableListener.java:418)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener.tryComplete(SubscribableListener.java:330)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener.setResult(SubscribableListener.java:355)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.SubscribableListener.onFailure(SubscribableListener.java:267)
	at org.elasticsearch.xpack.esql.plugin.ComputeService.lambda$executePlan$11(ComputeService.java:308)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:202)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:207)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onFailure(ActionListenerImplementations.java:360)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:263)
	at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$new$0(ComputeListener.java:37)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:233)
	at org.elasticsearch.compute.EsqlRefCountingListener.lambda$new$0(EsqlRefCountingListener.java:31)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:125)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:77)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.RefCountingRunnable.close(RefCountingRunnable.java:113)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.Releasables.close(Releasables.java:39)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.Releasables.closeExpectNoException(Releasables.java:72)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$2.run(ActionListenerImplementations.java:51)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:218)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:198)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:198)
	at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$acquireCompute$2(ComputeListener.java:63)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:258)
	at org.elasticsearch.xpack.esql.plugin.ComputeService.lambda$executePlan$19(ComputeService.java:419)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:233)
	at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$new$0(ComputeListener.java:37)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:233)
	at org.elasticsearch.compute.EsqlRefCountingListener.lambda$new$0(EsqlRefCountingListener.java:31)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:125)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:77)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.support.RefCountingRunnable.close(RefCountingRunnable.java:113)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.Releasables.close(Releasables.java:39)
	at org.elasticsearch.base@9.2.0-SNAPSHOT/org.elasticsearch.core.Releasables.closeExpectNoException(Releasables.java:72)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$2.run(ActionListenerImplementations.java:51)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:218)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:198)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:198)
	at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$acquireCompute$2(ComputeListener.java:63)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:258)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:350)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:111)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunAfterActionListener.onResponse(ActionListenerImplementations.java:312)
	at org.elasticsearch.compute.operator.DriverRunner$1.done(DriverRunner.java:65)
	at org.elasticsearch.compute.operator.DriverRunner$1.onResponse(DriverRunner.java:43)
	at org.elasticsearch.compute.operator.DriverRunner$1.onResponse(DriverRunner.java:40)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:258)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportResponseHandler$1.handleResponse(TransportResponseHandler.java:55)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportResponseHandler$Empty.handleResponse(TransportResponseHandler.java:45)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportResponseHandler$Empty.handleResponse(TransportResponseHandler.java:37)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleResponse(TransportService.java:1726)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1504)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:1602)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel$1.doRun(TransportService.java:1582)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1067)
	at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
	at org.elasticsearch.index.mapper.BlockSourceReader.toBytesRef(BlockSourceReader.java:473)
	at org.elasticsearch.index.mapper.BlockStoredFieldsReader$Id.read(BlockStoredFieldsReader.java:172)
	at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader$RowStrideReaderWork.read(ValuesFromSingleReader.java:216)
	at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromRowStrideReaders(ValuesFromSingleReader.java:168)
	at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromSingleLeaf(ValuesFromSingleReader.java:112)
	at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.load(ValuesFromSingleReader.java:58)
	at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:38)
	at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:18)
	at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:345)
	at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:321)
	at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:190)
	at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:175)
	at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator.getOutput(AbstractPageMappingToIteratorOperator.java:149)
	at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:272)
	at org.elasticsearch.compute.operator.Driver.run(Driver.java:186)
	at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:420)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.compute.operator.DriverScheduler$1.doRun(DriverScheduler.java:57)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1067)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.lang.Thread.run(Thread.java:1474)
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.index.mapper.BlockSourceReader.toBytesRef(BlockSourceReader.java:473)
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.index.mapper.BlockStoredFieldsReader$Id.read(BlockStoredFieldsReader.java:172)
		at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader$RowStrideReaderWork.read(ValuesFromSingleReader.java:216)
		at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromRowStrideReaders(ValuesFromSingleReader.java:168)
		at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromSingleLeaf(ValuesFromSingleReader.java:112)
		at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.load(ValuesFromSingleReader.java:58)
		at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:38)
		at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:18)
		at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:345)
		at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:321)
		at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:190)
		at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:175)
		at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator.getOutput(AbstractPageMappingToIteratorOperator.java:149)
		at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:272)
		at org.elasticsearch.compute.operator.Driver.run(Driver.java:186)
		at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:420)
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
		at org.elasticsearch.compute.operator.DriverScheduler$1.doRun(DriverScheduler.java:57)
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1067)
		at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
		at java.base/java.lang.Thread.run(Thread.java:1474)
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
	Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
		... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
		Suppressed: java.lang.NullPointerException: Cannot invoke "String.length()" because "v" is null
			... 25 more
```
</details>

--- a/docs/changelog/140102.yaml
+++ b/docs/changelog/140102.yaml
@@ -1,0 +1,7 @@
+pr: 140102
+summary: Add block loader for the `_id` field of time-series indices
+area: ES|QL
+type: bug
+issues:
+ - 140033
+ - 135689

--- a/server/src/main/java/org/elasticsearch/cluster/routing/RoutingHashBuilder.java
+++ b/server/src/main/java/org/elasticsearch/cluster/routing/RoutingHashBuilder.java
@@ -47,6 +47,10 @@ public class RoutingHashBuilder {
         }
     }
 
+    public void clear() {
+        hashes.clear();
+    }
+
     /**
      * Only expected to be called for old indices created before
      * {@link IndexVersions#TIME_SERIES_ROUTING_HASH_IN_ID} while creating (during ingestion)

--- a/server/src/main/java/org/elasticsearch/index/mapper/IdFieldMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/IdFieldMapper.java
@@ -127,7 +127,7 @@ public abstract class IdFieldMapper extends MetadataFieldMapper {
 
         @Override
         public BlockLoader blockLoader(BlockLoaderContext blContext) {
-            return new BlockStoredFieldsReader.IdBlockLoader();
+            return IdLoader.create(blContext.indexSettings(), blContext.mappingLookup()).blockLoader();
         }
 
         @Override

--- a/server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java
@@ -11,6 +11,7 @@ package org.elasticsearch.index.mapper;
 
 import org.apache.lucene.index.DocValues;
 import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.SortedDocValues;
 import org.apache.lucene.index.SortedNumericDocValues;
 import org.apache.lucene.index.SortedSetDocValues;
@@ -19,8 +20,10 @@ import org.elasticsearch.cluster.metadata.DataStream;
 import org.elasticsearch.cluster.routing.IndexRouting;
 import org.elasticsearch.cluster.routing.RoutingHashBuilder;
 import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.IndexVersions;
 import org.elasticsearch.index.fieldvisitor.LeafStoredFieldLoader;
+import org.elasticsearch.index.mapper.blockloader.docvalues.BlockDocValuesReader;
 
 import java.io.IOException;
 import java.util.ArrayList;
@@ -36,8 +39,7 @@ public sealed interface IdLoader permits IdLoader.TsIdLoader, IdLoader.StoredIdL
     /**
      * @return returns an {@link IdLoader} instance to load the value of the _id field.
      */
-    static IdLoader create(MapperService mapperService) {
-        var indexSettings = mapperService.getIndexSettings();
+    static IdLoader create(IndexSettings indexSettings, MappingLookup mappingLookup) {
         if (indexSettings.getMode() == IndexMode.TIME_SERIES) {
             IndexRouting.ExtractFromSource.ForRoutingPath indexRouting = null;
             List<String> routingPaths = null;
@@ -49,7 +51,7 @@ public sealed interface IdLoader permits IdLoader.TsIdLoader, IdLoader.StoredIdL
                         // In case the routing fields include path matches, find any matches and add them as distinct fields
                         // to the routing path.
                         Set<String> matchingRoutingPaths = new TreeSet<>(routingPaths);
-                        for (Mapper mapper : mapperService.mappingLookup().fieldMappers()) {
+                        for (Mapper mapper : mappingLookup.fieldMappers()) {
                             if (mapper instanceof KeywordFieldMapper && indexRouting.matchesField(mapper.fullPath())) {
                                 matchingRoutingPaths.add(mapper.fullPath());
                             }
@@ -84,6 +86,8 @@ public sealed interface IdLoader permits IdLoader.TsIdLoader, IdLoader.StoredIdL
     }
 
     Leaf leaf(LeafStoredFieldLoader loader, LeafReader reader, int[] docIdsInLeaf) throws IOException;
+
+    BlockLoader blockLoader();
 
     /**
      * Returns a leaf instance for a leaf reader that returns the _id for segment level doc ids.
@@ -171,15 +175,163 @@ public sealed interface IdLoader permits IdLoader.TsIdLoader, IdLoader.StoredIdL
             }
             return new TsIdLeaf(docIdsInLeaf, ids);
         }
+
+        @Override
+        public BlockLoader blockLoader() {
+            return new BlockDocValuesReader.DocValuesBlockLoader() {
+                @Override
+                public AllReader reader(LeafReaderContext context) throws IOException {
+                    if (indexRouting != null) {
+                        return new LegacyTsIdFieldReader(context.reader(), indexRouting, routingPaths);
+                    } else {
+                        return new TsIdFieldReader(context.reader(), useSyntheticId);
+                    }
+                }
+
+                @Override
+                public Builder builder(BlockFactory factory, int expectedCount) {
+                    return factory.bytesRefs(expectedCount);
+                }
+            };
+        }
+
+        private static class TsIdFieldReader extends BlockDocValuesReader {
+            final SortedDocValues tsidDVs;
+            final SortedNumericDocValues timestampDVs;
+            final SortedDocValues routingHashDVs;
+            final boolean useSyntheticId;
+
+            TsIdFieldReader(LeafReader leafReader, boolean useSyntheticId) throws IOException {
+                this.tsidDVs = DocValues.getSorted(leafReader, TimeSeriesIdFieldMapper.NAME);
+                this.timestampDVs = DocValues.getSortedNumeric(leafReader, DataStream.TIMESTAMP_FIELD_NAME);
+                this.routingHashDVs = DocValues.getSorted(leafReader, TimeSeriesRoutingHashFieldMapper.NAME);
+                this.useSyntheticId = useSyntheticId;
+            }
+
+            @Override
+            protected int docId() {
+                return tsidDVs.docID();
+            }
+
+            @Override
+            public String toString() {
+                return "TsIdFieldReader";
+            }
+
+            @Override
+            public BlockLoader.Block read(BlockLoader.BlockFactory factory, BlockLoader.Docs docs, int offset, boolean nullsFiltered)
+                throws IOException {
+                try (var builder = factory.bytesRefs(docs.count() - offset)) {
+                    for (int i = offset; i < docs.count(); i++) {
+                        read(docs.get(i), null, builder);
+                    }
+                    return builder.build();
+                }
+            }
+
+            @Override
+            public void read(int docId, BlockLoader.StoredFields storedFields, BlockLoader.Builder builder) throws IOException {
+                if (tsidDVs.advanceExact(docId) == false
+                    || timestampDVs.advanceExact(docId) == false
+                    || routingHashDVs.advanceExact(docId) == false) {
+                    assert false : "_tsid or @timestamp or _ts_routing_hash missing for docId " + docId;
+                    throw new IllegalStateException("_tsid or @timestamp or _ts_routing_hash missing for docId " + docId);
+                }
+                BytesRef tsid = tsidDVs.lookupOrd(tsidDVs.ordValue());
+                long timestamp = timestampDVs.nextValue();
+                BytesRef routingHashBytes = routingHashDVs.lookupOrd(routingHashDVs.ordValue());
+                int routingHash = TimeSeriesRoutingHashFieldMapper.decode(
+                    Uid.decodeId(routingHashBytes.bytes, routingHashBytes.offset, routingHashBytes.length)
+                );
+                final String id;
+                if (useSyntheticId) {
+                    id = TsidExtractingIdFieldMapper.createSyntheticId(tsid, timestamp, routingHash);
+                } else {
+                    id = TsidExtractingIdFieldMapper.createId(routingHash, tsid, timestamp);
+                }
+                ((BlockLoader.BytesRefBuilder) builder).appendBytesRef(new BytesRef(id));
+            }
+        }
+
+        private static class LegacyTsIdFieldReader extends BlockDocValuesReader {
+            final RoutingHashBuilder routingBuilder;
+            final SortedDocValues tsidDVs;
+            final SortedNumericDocValues timestampDVs;
+            final SortedDocValues[] routingHashDVs;
+            final List<String> routingPaths;
+            final byte[] scratch = new byte[16];
+
+            LegacyTsIdFieldReader(
+                LeafReader leafReader,
+                IndexRouting.ExtractFromSource.ForRoutingPath indexRouting,
+                List<String> routingPaths
+            ) throws IOException {
+                this.routingBuilder = indexRouting.builder();
+                this.routingPaths = routingPaths;
+                this.routingHashDVs = new SortedDocValues[routingPaths.size()];
+                for (int i = 0; i < routingPaths.size(); i++) {
+                    this.routingHashDVs[i] = DocValues.getSorted(leafReader, routingPaths.get(i));
+                }
+                this.tsidDVs = DocValues.getSorted(leafReader, TimeSeriesIdFieldMapper.NAME);
+                this.timestampDVs = DocValues.getSortedNumeric(leafReader, DataStream.TIMESTAMP_FIELD_NAME);
+            }
+
+            @Override
+            protected int docId() {
+                return tsidDVs.docID();
+            }
+
+            @Override
+            public String toString() {
+                return "LegacyTsIdFieldReader";
+            }
+
+            @Override
+            public BlockLoader.Block read(BlockLoader.BlockFactory factory, BlockLoader.Docs docs, int offset, boolean nullsFiltered)
+                throws IOException {
+                try (var builder = factory.bytesRefs(docs.count() - offset)) {
+                    for (int i = offset; i < docs.count(); i++) {
+                        read(docs.get(i), null, builder);
+                    }
+                    return builder.build();
+                }
+            }
+
+            @Override
+            public void read(int docId, BlockLoader.StoredFields storedFields, BlockLoader.Builder builder) throws IOException {
+                if (tsidDVs.advanceExact(docId) == false || timestampDVs.advanceExact(docId) == false) {
+                    assert false : "_tsid or @timestamp missing for docId " + docId;
+                    throw new IllegalStateException("_tsid or @timestamp missing for docId " + docId);
+                }
+                routingBuilder.clear();
+                BytesRef tsid = tsidDVs.lookupOrd(tsidDVs.ordValue());
+                long timestamp = timestampDVs.nextValue();
+                for (int i = 0; i < routingHashDVs.length; i++) {
+                    SortedDocValues dv = routingHashDVs[i];
+                    if (dv.advanceExact(docId)) {
+                        BytesRef v = dv.lookupOrd(dv.ordValue());
+                        routingBuilder.addMatching(routingPaths.get(i), v);
+                    }
+                }
+                var id = TsidExtractingIdFieldMapper.createId(false, routingBuilder, tsid, timestamp, scratch);
+                ((BlockLoader.BytesRefBuilder) builder).appendBytesRef(new BytesRef(id));
+            }
+        }
     }
 
     final class StoredIdLoader implements IdLoader {
+        public StoredIdLoader() {
 
-        public StoredIdLoader() {}
+        }
 
         @Override
         public Leaf leaf(LeafStoredFieldLoader loader, LeafReader reader, int[] docIdsInLeaf) throws IOException {
             return new StoredLeaf(loader);
+        }
+
+        @Override
+        public BlockLoader blockLoader() {
+            return new BlockStoredFieldsReader.IdBlockLoader();
         }
     }
 

--- a/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
+++ b/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
@@ -38,6 +38,7 @@ import org.elasticsearch.index.fielddata.IndexNumericFieldData;
 import org.elasticsearch.index.fielddata.IndexOrdinalsFieldData;
 import org.elasticsearch.index.mapper.IdLoader;
 import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.MapperService;
 import org.elasticsearch.index.mapper.NestedLookup;
 import org.elasticsearch.index.mapper.SourceLoader;
 import org.elasticsearch.index.query.AbstractQueryBuilder;
@@ -946,6 +947,7 @@ final class DefaultSearchContext extends SearchContext {
 
     @Override
     public IdLoader newIdLoader() {
-        return IdLoader.create(indexService.mapperService());
+        MapperService mapperService = indexService.mapperService();
+        return IdLoader.create(mapperService.getIndexSettings(), mapperService.mappingLookup());
     }
 }

--- a/server/src/test/java/org/elasticsearch/index/mapper/TsidExtractingIdFieldMapperTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/TsidExtractingIdFieldMapperTests.java
@@ -11,6 +11,7 @@ package org.elasticsearch.index.mapper;
 import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
 
 import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.util.BytesRef;
 import org.elasticsearch.cluster.metadata.IndexMetadata;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.settings.Settings;
@@ -758,39 +759,52 @@ public class TsidExtractingIdFieldMapperTests extends MetadataMapperTestCase {
     }
 
     private final TestCase testCase;
+    private final BlockLoaderTestRunner blockLoaderTestRunner;
 
     private static final int ROUTING_HASH = 7;
 
     public TsidExtractingIdFieldMapperTests(@Named("testCase") TestCase testCase) {
         this.testCase = testCase;
+        this.blockLoaderTestRunner = new BlockLoaderTestRunner(
+            new BlockLoaderTestCase.Params(false, randomFrom(MappedFieldType.FieldExtractPreference.values())),
+            false
+        );
     }
 
     public void testExpectedIdWithRoutingPath() throws IOException {
-        assertThat(parse(mapperService(false), testCase.source).id(), equalTo(testCase.expectedIdWithRoutingPath));
+        MapperService mapperService = mapperService(false);
+        ParsedDocument parsed = parse(mapperService, testCase.source);
+        assertThat(parsed.id(), equalTo(testCase.expectedIdWithRoutingPath));
+        verifyIdFromBlockLoader(mapperService, parsed, testCase.expectedIdWithRoutingPath);
     }
 
     public void testExpectedIdWithIndexDimensions() throws IOException {
-        assertThat(parse(mapperService(true), testCase.source).id(), equalTo(testCase.expectedIdWithIndexDimensions));
+        MapperService mapperService = mapperService(true);
+        ParsedDocument parsed = parse(mapperService, testCase.source);
+        assertThat(parsed.id(), equalTo(testCase.expectedIdWithIndexDimensions));
+        verifyIdFromBlockLoader(mapperService, parsed, testCase.expectedIdWithIndexDimensions);
     }
 
     public void testProvideExpectedIdWithRoutingPath() throws IOException {
-        assertThat(
-            parse(testCase.expectedIdWithRoutingPath, mapperService(false), testCase.source).id(),
-            equalTo(testCase.expectedIdWithRoutingPath)
-        );
+        MapperService mapperService = mapperService(false);
+        ParsedDocument parsed = parse(testCase.expectedIdWithRoutingPath, mapperService, testCase.source);
+        assertThat(parsed.id(), equalTo(testCase.expectedIdWithRoutingPath));
+        verifyIdFromBlockLoader(mapperService, parsed, testCase.expectedIdWithRoutingPath);
     }
 
     public void testProvideExpectedIdWithIndexDimensions() throws IOException {
-        assertThat(
-            parse(testCase.expectedIdWithIndexDimensions, mapperService(true), testCase.source).id(),
-            equalTo(testCase.expectedIdWithIndexDimensions)
-        );
+        MapperService mapperService = mapperService(true);
+        ParsedDocument parsed = parse(testCase.expectedIdWithIndexDimensions, mapperService, testCase.source);
+        assertThat(parsed.id(), equalTo(testCase.expectedIdWithIndexDimensions));
+        verifyIdFromBlockLoader(mapperService, parsed, testCase.expectedIdWithIndexDimensions);
     }
 
     public void testEquivalentSourcesWithRoutingPath() throws IOException {
         MapperService mapperService = mapperService(false);
         for (CheckedConsumer<XContentBuilder, IOException> equivalent : testCase.equivalentSources) {
-            assertThat(parse(mapperService, equivalent).id(), equalTo(testCase.expectedIdWithRoutingPath));
+            ParsedDocument parsed = parse(mapperService, equivalent);
+            assertThat(parsed.id(), equalTo(testCase.expectedIdWithRoutingPath));
+            verifyIdFromBlockLoader(mapperService, parsed, testCase.expectedIdWithRoutingPath);
         }
     }
 
@@ -975,5 +989,9 @@ public class TsidExtractingIdFieldMapperTests extends MetadataMapperTestCase {
                     + "]"
             )
         );
+    }
+
+    private void verifyIdFromBlockLoader(MapperService mapperService, ParsedDocument doc, String expectedId) throws IOException {
+        blockLoaderTestRunner.runTest(mapperService, doc, new BytesRef(expectedId), "_id");
     }
 }

--- a/server/src/test/java/org/elasticsearch/index/mapper/blockloader/TextFieldWithParentBlockLoaderTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/blockloader/TextFieldWithParentBlockLoaderTests.java
@@ -40,7 +40,7 @@ public class TextFieldWithParentBlockLoaderTests extends MapperServiceTestCase {
 
     public TextFieldWithParentBlockLoaderTests(BlockLoaderTestCase.Params params) {
         this.params = params;
-        this.runner = new BlockLoaderTestRunner(params);
+        this.runner = new BlockLoaderTestRunner(params, randomBoolean());
     }
 
     // This is similar to BlockLoaderTestCase#testBlockLoaderOfMultiField but has customizations required to properly test the case

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestCase.java
@@ -65,7 +65,7 @@ public abstract class BlockLoaderTestCase extends MapperServiceTestCase {
         this.fieldType = fieldType;
         this.params = params;
         this.customDataSourceHandlers = customDataSourceHandlers;
-        this.runner = new BlockLoaderTestRunner(params);
+        this.runner = new BlockLoaderTestRunner(params, randomBoolean());
 
         this.fieldName = randomAlphaOfLengthBetween(5, 10);
     }

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestRunner.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestRunner.java
@@ -45,7 +45,6 @@ public class BlockLoaderTestRunner {
     private final BlockLoaderTestCase.Params params;
     private final boolean allowDummyDocs;
 
-
     public BlockLoaderTestRunner(BlockLoaderTestCase.Params params, boolean allowDummyDocs) {
         this.params = params;
         this.allowDummyDocs = allowDummyDocs;

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestRunner.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/BlockLoaderTestRunner.java
@@ -43,35 +43,43 @@ import static org.hamcrest.Matchers.equalTo;
 
 public class BlockLoaderTestRunner {
     private final BlockLoaderTestCase.Params params;
+    private final boolean allowDummyDocs;
 
-    public BlockLoaderTestRunner(BlockLoaderTestCase.Params params) {
+
+    public BlockLoaderTestRunner(BlockLoaderTestCase.Params params, boolean allowDummyDocs) {
         this.params = params;
+        this.allowDummyDocs = allowDummyDocs;
     }
 
     public void runTest(MapperService mapperService, Map<String, Object> document, Object expected, String blockLoaderFieldName)
         throws IOException {
         var documentXContent = XContentBuilder.builder(XContentType.JSON.xContent()).map(document);
+        var source = new SourceToParse(
+            "1",
+            BytesReference.bytes(documentXContent),
+            XContentType.JSON,
+            null,
+            Map.of(),
+            Map.of(),
+            true,
+            XContentMeteringParserDecorator.NOOP,
+            null
+        );
+        var parsedDoc = mapperService.documentMapper().parse(source);
+        runTest(mapperService, parsedDoc, expected, blockLoaderFieldName);
+    }
 
-        Object blockLoaderResult = setupAndInvokeBlockLoader(mapperService, documentXContent, blockLoaderFieldName);
+    public void runTest(MapperService mapperService, ParsedDocument parsedDoc, Object expected, String blockLoaderFieldName)
+        throws IOException {
+        Object blockLoaderResult = setupAndInvokeBlockLoader(mapperService, parsedDoc, blockLoaderFieldName);
         assertThat(blockLoaderResult, prettyEqualTo(expected));
     }
 
-    private Object setupAndInvokeBlockLoader(MapperService mapperService, XContentBuilder document, String fieldName) throws IOException {
+    private Object setupAndInvokeBlockLoader(MapperService mapperService, ParsedDocument parsedDoc, String fieldName) throws IOException {
         try (Directory directory = newDirectory()) {
             RandomIndexWriter iw = new RandomIndexWriter(random(), directory);
 
-            var source = new SourceToParse(
-                "1",
-                BytesReference.bytes(document),
-                XContentType.JSON,
-                null,
-                Map.of(),
-                Map.of(),
-                true,
-                XContentMeteringParserDecorator.NOOP,
-                null
-            );
-            LuceneDocument doc = mapperService.documentMapper().parse(source).rootDoc();
+            LuceneDocument doc = parsedDoc.rootDoc();
 
             /*
              * Add three documents with doc id 0, 1, 2. The real document is 1.
@@ -110,7 +118,7 @@ public class BlockLoaderTestRunner {
                     } else if (i == offset) {
                         docArray[i] = 1;
                     } else {
-                        docArray[i] = 2;
+                        docArray[i] = allowDummyDocs ? 2 : 1;
                     }
                 }
             }

--- a/x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/TimeSeriesIT.java
+++ b/x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/TimeSeriesIT.java
@@ -22,6 +22,8 @@ import org.elasticsearch.compute.operator.OperatorStatus;
 import org.elasticsearch.compute.operator.TimeSeriesAggregationOperator;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.core.Tuple;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.xpack.esql.EsqlTestUtils;
 import org.elasticsearch.xpack.esql.core.type.DataType;
 import org.elasticsearch.xpack.esql.plugin.QueryPragmas;
@@ -108,7 +110,10 @@ public class TimeSeriesIT extends AbstractEsqlIntegTestCase {
 
     @Before
     public void populateIndex() {
-        Settings settings = Settings.builder().put("mode", "time_series").putList("routing_path", List.of("host", "cluster")).build();
+        Settings.Builder settings = Settings.builder().put("mode", "time_series").putList("routing_path", List.of("host", "cluster"));
+        if (IndexSettings.TSDB_SYNTHETIC_ID_FEATURE_FLAG && randomBoolean()) {
+            settings.put("index.codec", "default").put("index.number_of_replicas", 0).put("index.mapping.use_synthetic_id", true);
+        }
         client().admin()
             .indices()
             .prepareCreate("hosts")
@@ -981,6 +986,34 @@ public class TimeSeriesIT extends AbstractEsqlIntegTestCase {
                 assertThat(sumOfAvgPerTBucket.get(tbucket), closeTo(expectedSumOfAvgPerTBucket.get(tbucket), 0.5));
             }
         }
+    }
+
+    public void testLoadId() {
+        Map<String, Tuple<String, String>> fromEsql = new HashMap<>();
+        try (var resp = run("FROM hosts* METADATA _id, _tsid | KEEP _id, _tsid, @timestamp | LIMIT 1000")) {
+            List<List<Object>> rows = EsqlTestUtils.getValuesList(resp);
+            for (List<Object> row : rows) {
+                String id = row.get(0).toString();
+                String tsid = row.get(1).toString();
+                String timestamp = row.get(2).toString();
+                fromEsql.put(id, Tuple.tuple(tsid, timestamp));
+            }
+        }
+        var searchResponse = client().prepareSearch("hosts*")
+            .addFetchField("_tsid")
+            .addFetchField("@timestamp")
+            .setSize(1000)
+            .setFetchSource(false)
+            .get();
+        Map<String, Tuple<String, String>> fromSearch = new HashMap<>();
+        for (SearchHit hit : searchResponse.getHits().getHits()) {
+            String id = hit.getId();
+            String tsid = hit.getFields().get("_tsid").getValue().toString();
+            String timestamp = hit.getFields().get("@timestamp").getValue().toString();
+            fromSearch.put(id, Tuple.tuple(tsid, timestamp));
+        }
+        searchResponse.decRef();
+        assertThat(fromEsql, equalTo(fromSearch));
     }
 
     private static double round(double value) {

